pkg/backup/snapshotter/snapshotter.go

```diff
@@ -475,20 +484,85 @@ func (m *StoresMixture) ProcessCSBPVCsAndPVs(r *v1alpha1.Restore, csb *CloudSnap
 		// Reset the PV's binding status so that Kubernetes can properly
 		// associate it with the restored PVC.
 		resetVolumeBindingInfo(pvc, pv)
-
 		// Reset the PV's volumeID for restore from snapshot
-		// Note: This function has to precede the following `resetMetadataAndStatus`
-		if reason, err := m.resetRestoreVolumeID(pv); err != nil {
-			return reason, err
+		if err := m.snapshotter.SetVolumeID(pv, restoreVolID); err != nil {
+			return "ResetRestoreVolumeIDFailed", fmt.Errorf("failed to set pv-%s, %s", pv.Name, err.Error())
 		}
-
 		// Clear out non-core metadata fields and status
 		resetMetadataAndStatus(r, backupClusterName, pvc, pv)
+
+		pvs = append(pvs, pv)
+		pvcs = append(pvcs, pvc)
+	}
+
+	restoreSTSName := controller.TiKVMemberName(r.Spec.BR.Cluster)
+	sequentialPVCs, sequentialPVs, err := resetPVCSequence(restoreSTSName, pvcs, pvs)
+	if err != nil {
+		klog.Errorf("reset pvcs to sequential error: %s", err.Error())
+		return "InvalidPVCName", err
 	}
 
+	csb.Kubernetes.PVCs = sequentialPVCs
+	csb.Kubernetes.PVs = sequentialPVs
 	return "", nil
 }
 
+// If a backup cluster has scaled in and the tidb-operator enabled advanced-statefulset(ref: https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/advanced-statefulset)
+// the name of pvc can be not sequential, eg: [pvc-0, pvc-1, pvc-2, pvc-6, pvc-7, pvc-8]
+// When create cluster, we should ensure pvc name sequential, so we should reset it to [pvc-0, pvc-1, pvc-2, pvc-3, pvc-4, pvc-5]
+func resetPVCSequence(stsName string, pvcs []*corev1.PersistentVolumeClaim, pvs []*corev1.PersistentVolume) (
+	[]*corev1.PersistentVolumeClaim, []*corev1.PersistentVolume, error) {
+	type indexedPVC struct {
+		index int
+		pvc   *corev1.PersistentVolumeClaim
+	}
+	indexedPVCs := make([]*indexedPVC, 0, len(pvcs))
+	reStr := fmt.Sprintf(`%s-(\d+)$`, stsName)
+	re := regexp.MustCompile(reStr)
+	for _, pvc := range pvcs {
+		subMatches := re.FindStringSubmatch(pvc.Name)
+		if len(subMatches) != 2 {
+			return nil, nil, fmt.Errorf("pvc name %s doesn't match regex %s", pvc.Name, reStr)
+		}
+		index, err := strconv.Atoi(subMatches[1])
+		if err != nil {
+			return nil, nil, fmt.Errorf("parse index %s of pvc %s to int: %s", subMatches[1], pvc.Name, err.Error())
+		}
+		indexedPVCs = append(indexedPVCs, &indexedPVC{
+			index: index,
+			pvc:   pvc,
+		})
+	}
+
+	sort.Slice(indexedPVCs, func(i, j int) bool {
+		return indexedPVCs[i].index < indexedPVCs[j].index
+	})
+	pvc2pv := make(map[string]*corev1.PersistentVolume, len(pvs))
+	for _, pv := range pvs {
+		pvc2pv[pv.Spec.ClaimRef.Name] = pv
+	}
+
+	sequentialPVCs := make([]*corev1.PersistentVolumeClaim, 0, len(pvcs))
+	sequentialPVs := make([]*corev1.PersistentVolume, 0, len(pvs))
+	for i, iPVC := range indexedPVCs {
+		pv, ok := pvc2pv[iPVC.pvc.Name]
+		if !ok {
+			return nil, nil, fmt.Errorf("pv with claim %s not found", iPVC.pvc.Name)
+		}
+		if i != iPVC.index {
```

Comment:
some comments to explain why this happen?

Reply:
e.g: PingCAP advance sts introduce a new delete strategy, delete-slots allow to deletion of arbitrary instances specify by numbers [3.4.5]. this causes the index between pvc and pv may not to be equal during the restore.

Reply:
done